more_itertools/more.py

```diff
@@ -949,7 +1105,35 @@ def walk(node, level):
     yield from walk(iterable, 0)
 
 
-def side_effect(func, iterable, chunk_size=None, before=None, after=None):
+@overload
+def side_effect(
+    func: Callable[[_T], object],
+    iterable: Iterable[_T],
+    chunk_size: None = ...,
+    before: Optional[Callable[[], object]] = ...,
+    after: Optional[Callable[[], object]] = ...,
+) -> Iterator[_T]:
+    ...
+
+
+@overload
+def side_effect(
+    func: Callable[[List[_T]], object],
+    iterable: Iterable[_T],
+    chunk_size: int,
+    before: Optional[Callable[[], object]] = ...,
+    after: Optional[Callable[[], object]] = ...,
+) -> Iterator[_T]:
+    ...
+
+
+def side_effect(
+    func: Union[Callable[[_T], object], Callable[[List[_T]], object]],
+    iterable: Iterable[_T],
+    chunk_size: Optional[int] = None,
+    before: Optional[Callable[[], object]] = None,
```

Comment:
Should this be `Optional[Callable[], Any]`, judging from the example of callbacks in https://www.python.org/dev/peps/pep-0484/#the-any-type ?

Reply:
`Any` and `object` are similar but slightly different. `Callable[[], object]` will be stricter because of this:

> When the type of a value is `object`, the type checker will reject almost all operations on it, and assigning it to a variable (or using it as a return value) of a more specialized type is a type error. On the other hand, when a value has type `Any`, the type checker will allow all operations on it, and a value of type `Any` can be assigned to a variable (or used as a return value) of a more constrained type.

If we take a function as an argument and call it, we can only use the return value in a way that would be compatible with any return type. For instance, we shouldn't be able to add an integer to the result, because the result might be of type `List[complex]` for instance.

```python
a: Callable[[], Any]
b: Callable[[], object]

if a():
    ...
if b():
    ...

a() + 5
b() + 5  #  error: Unsupported operand types for + ("object" and "int")  [operator]
```

Reply:
Got it! Thanks for spelling that out for me.